### PR TITLE
Java support for explode position

### PR DESCRIPTION
**Pull request by sperlingxx**
This pull request provides Java interface for `Table.explode_position`, which is required by spark-rapids plugin.

### CODE REVIEW

**Review by vuule — APPROVED** [review]

**Review by cwharris — APPROVED** [review]

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 1]
```diff
@@ -1753,6 +1755,48 @@ public Table explode(int index) {
...
+   * [null,        200],
+   * [[],          300],
+   * returns
+   * [0,    5,     100],
+   * [1,    null,  100],
+   * [2,    15,    100],
+   * </code>
+   *
+   * Note that null lists are not included in the resulting table, but nulls inside
+   * lists and empty lists will be represented with a null entry for that column in that row.
```
Comment (hyperbolic2346) [inline java/src/main/java/ai/rapids/cudf/Table.java:1789]: I know this is a copy of the comment from cudf, but is this actually true? I don't think that empty lists will be exposed as a null entry for the column and the block above shows the empty list disappearing as well. I believe empty lists not being represented in the output is the desired behavior, but I'd like to hear from @revans2 or @jlowe to be certain.

**Review by hyperbolic2346 — CHANGES_REQUESTED** [review]
Looks good, thanks for the comment fix on column ordering. Once we get verification on empty lists we can update some comments and move forward.

File: java/src/main/java/ai/rapids/cudf/Table.java (hunk 1 above)
Comment (jlowe) [reply java/src/main/java/ai/rapids/cudf/Table.java:1789]: Spark does not discard null entries nor empty lists, e.g.:
```
scala> sql("select posexplode(array(array(5,null,15), null, array()))").show
+---+--------+
|pos|     col|
+---+--------+
|  0|[5,, 15]|
|  1|    null|
|  2|      []|
+---+--------+


scala> sql("select posexplode(array(array(5,null,15), null, array()))").collect
res20: Array[org.apache.spark.sql.Row] = Array([0,WrappedArray(5, null, 15)], [1,null], [2,WrappedArray()])
```

Comment (jlowe) [reply java/src/main/java/ai/rapids/cudf/Table.java:1789]: Sorry, that was posexplode in isolation on a column.  When applied to a table with other columns the semantics indeed match Spark:
```
>>> df.show()
+--------+---+
|     arr|val|
+--------+---+
|[5,, 15]|100|
|    null|200|
|      []|300|
+--------+---+
>>> df.select(posexplode(df.arr),df.val).show()
+---+----+---+
|pos| col|val|
+---+----+---+
|  0|   5|100|
|  1|null|100|
|  2|  15|100|
+---+----+---+
```

**Review by jlowe — APPROVED** [review]

File: java/src/main/java/ai/rapids/cudf/Table.java (hunk 1 above)
Comment (sperlingxx) [reply java/src/main/java/ai/rapids/cudf/Table.java:1789]: I agree with Jason that this comment is correct. And I add some null/empty entities into test to ensure cuDF working as the comment describing.

Comment (hyperbolic2346) [reply java/src/main/java/ai/rapids/cudf/Table.java:1789]: I may have been unclear. The issue I see is the wording `empty lists will be represented with a null entry for that column in that row` as an empty list is not represented in the output as there is no 300 in this example. My suggestion would be to change the comment to  say that:
```
* Note that null and empty lists are not included in the resulting table, but nulls inside
* lists will be represented with a null entry for that column in that row.
```

File: java/src/test/java/ai/rapids/cudf/TableTest.java [hunk 2]
```diff
@@ -4550,10 +4550,20 @@ void testExplode() {
             .column("s1", "s1", "s1", "s2", "s2", "s3")
             .column(   1,    1,    1,    3,    3,    5)
             .column(12.0, 12.0, 12.0, 14.0, 14.0, 13.0)
-            .build()) {
+            .build();
+         Table expectedPos = new Table.TestBuilder()
```
[the remark this answers is not in the document]
Comment (sperlingxx) [reply java/src/test/java/ai/rapids/cudf/TableTest.java]: It's done.

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 3]
```diff
@@ -1753,6 +1755,48 @@ public Table explode(int index) {
...
+   *
+   * @param index Column index to explode inside the table.
+   * @return A new table with explode_col exploded.
```
[the remark this answers is not in the document]
Comment (sperlingxx) [reply java/src/main/java/ai/rapids/cudf/Table.java]: Yes, I added column order info.

File: cpp/include/cudf/reshape.hpp [hunk 4]
```diff
@@ -150,12 +150,12 @@ std::unique_ptr<table> explode(
...
- * [15,  2,    100],
- * [20,  0,    200],
- * [25,  1,    200],
- * [30,  0,    300],
+ * [0,   5,    100],
+ * [1,   10,   100],
+ * [2,   15,    100],
+ * [0,   20,    200],
+ * [1,   25,    200],
+ * [0,   30,    300],
```
[the remark this answers is not in the document]
Comment (hyperbolic2346) [reply cpp/include/cudf/reshape.hpp:158]: Yes, this ordering is incorrect in cudf. Thanks for finding this and updating it, @sperlingxx.

**Review by hyperbolic2346 — APPROVED** [review]

**Comment by sperlingxx** [comment]
build

**Review by revans2 — APPROVED** [review]

**Comment by revans2** [comment]
Because this is just a comment change to the C++ code, is it okay if we skip ci on this to merge it in?

**Comment by hyperbolic2346** [comment]
I would think so. We could also pull the comment changes into a different PR if they were going to hold this one up.

**Comment by kkraus14** [comment]
I'm fine with marking this as `[skip ci]`

**Comment by cwharris** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by sperlingxx** [comment]
@gpucibot merge

**Pull request merged**